### PR TITLE
Fix array_sum return type

### PR DESCRIPTION
**Pull request by VincentLanglet**

### CODE REVIEW

File: src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -36,12 +36,20 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 		}
 
 		if ($arrayType->isIterableAtLeastOnce()->yes()) {
-			return TypeCombinator::intersect($itemType, new UnionType([new IntegerType(), new FloatType()]));
+			if ($itemType instanceof IntegerType || $itemType instanceof FloatType) {
```
Comment (ondrejmirtes) [inline src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php:39]: Better comparison would be:

`(new UnionType([new IntegerType(), new FloatType()]))->isSuperTypeOf($itemType)->yes()`

File: src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php [hunk 2]
```diff
@@ -36,12 +36,20 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
...
+				return $itemType;
+			}
+
+			return new UnionType([new IntegerType(), new FloatType()]);
+		}
+
+		if ($itemType instanceof IntegerType || $itemType instanceof FloatType) {
```
Comment (ondrejmirtes) [inline src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php:46]: dtto

File: src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php [hunk 3]
```diff
@@ -36,12 +36,20 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
...
+			return TypeCombinator::union(new ConstantIntegerType(0), $itemType);
 		}
 
 		return TypeCombinator::union(
 			new ConstantIntegerType(0),
-			TypeCombinator::intersect($itemType, new UnionType([new IntegerType(), new FloatType()]))
+			new UnionType([new IntegerType(), new FloatType()])
```
Comment (ondrejmirtes) [inline src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php:52]: I don't think it makes sense to modify this.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php (hunk 3 above)
Comment (VincentLanglet) [reply src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php:52]: For an array of string, $itemType is a stringType.
So `TypeCombinator::intersect($itemType, new UnionType([new IntegerType(), new FloatType()])` is nothing.
That's why the dynamic return type return  `new ConstantIntegerType(0)`.

The intersect seems to be the reason of the bug.

Comment (ondrejmirtes) [reply src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php]: But on the other hand, this intersect allows us to specify that when we only have `array<int>`, the result is going to be `int`.

Comment (VincentLanglet) [reply src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php]: Isn't this handle by the previous check
```
if ($intUnionFloat->isSuperTypeOf($itemType)->yes()) {
    return TypeCombinator::union(new ConstantIntegerType(0), $itemType);
}
```
?

**Comment by ondrejmirtes** [comment]
Awesome! 👍

**Review by ondrejmirtes — APPROVED** [review]

**Pull request merged**